python/ray/data/_internal/execution/operators/hash_shuffle.py

```diff
@@ -445,12 +446,152 @@ def __init__(
             int, Dict[int, _PartitionStats]
         ] = defaultdict(dict)
 
+        # Resource monitoring state
+        self._aggregator_startup_time: Optional[float] = None
+
+        # Add last warning timestamp for health checks
+        self._last_health_warning_time: Optional[float] = None
+        self._health_warning_interval: float = 10.0  # 10 seconds between warnings
+        # Track readiness refs for non-blocking health checks
+        self._readiness_refs: Optional[List[ObjectRef]] = None
+
     def start(self, options: ExecutionOptions) -> None:
         super().start(options)
 
+        # Check cluster resources before starting aggregators
+        self._check_cluster_resources()
+
+        # Record start time for monitoring
+        self._aggregator_startup_time = time.time()
+
         self._aggregator_pool.start()
 
+    def _check_cluster_resources(self) -> None:
+        """Check if cluster has enough resources to schedule all aggregators.
+
+        Raises:
+            ValueError: If cluster doesn't have sufficient resources.
+        """
+        try:
+            cluster_resources = ray.cluster_resources()
+            available_resources = ray.available_resources()
+        except Exception as e:
+            logger.warning(f"Failed to get cluster resources: {e}")
+            return
+
+        # Calculate required resources for all aggregators
+        aggregator_args = self._aggregator_pool._aggregator_ray_remote_args
+        required_cpus = (
+            aggregator_args.get("num_cpus", 1) * self._aggregator_pool.num_aggregators
+        )
+        required_memory = (
+            aggregator_args.get("memory", 0) * self._aggregator_pool.num_aggregators
+        )
+
+        # Check CPU resources
+        total_cpus = cluster_resources.get("CPU", 0)
+        available_cpus = available_resources.get("CPU", 0)
+
+        if required_cpus > total_cpus:
+            raise ValueError(
+                f"Insufficient CPU resources in cluster for hash shuffle operation. "
+                f"Required: {required_cpus} CPUs for {self._aggregator_pool.num_aggregators} aggregators, "
+                f"but cluster only has {total_cpus} total CPUs. "
+                f"Consider either increasing the cluster size or reducing the number of aggregators via `DataContext.max_hash_shuffle_aggregators`."
+            )
```

Comment:
I've just realized that this branch would fail for autoscaling clusters:

 - We can't throw an exception b/c of that
 - We'd instead print a warning mentioning that we recommend starting from a bigger cluster (to accommodate Actor Pool